### PR TITLE
Added Zfinx and Zfh tests

### PR DESCRIPTION
**Pull request by anuani21**
<FOR DOC UPDATES FILL ONLY DESCRIPTION AND RELATED ISSUES SECTION AND REMOVE THE OTHERS>

## Description

-   Zfh and Zfinx Test are added
- Added half width and half word into the NAN function to accommodate Zfh, as the base for Zfh is F which is also falls under the value as word
- Introduced two new flags (ZFINX and ZDINX) to aid the compilation of floating ops into integer regs

### Related Issues
 NA

### Ratified/Unratified Extensions

[✓] Ratified

### List Extensions

- Zfinx
- Zfh

### Reference Model Used

[✓] SAIL

### Mandatory Checklist:

Ran reports are placed here

https://gitlab.com/ptprasanna/actreports.git

### CODE REVIEW

**Comment by pawks** [comment]
As already discussed, this PR can be merged only after the corresponding PR for ISAC has been merged.

**Comment by davidharrishmc** [comment]
I would use these tests on CORE-V Wally when they are available.

PR72 for ISAC has been merged.
https://github.com/riscv-software-src/riscv-isac/pull/72

PR67 from CTG has been waiting for review since July 2.
https://github.com/riscv-software-src/riscv-ctg/pull/67

**Comment by anuani21** [comment]
@neelgala , -Please have your review and let us know if more information would help.

**Comment by davidharrishmc** [comment]
I see there are no fma tests in this directory.  Is there a reason these are omitted?

**Comment by anuani21** [comment]
Hi,

As per the Zfinx and Zfh specification, there is no fma tests available I
think

**Comment by davidharrishmc** [comment]
Got it. Thanks!

**Comment by davidharrishmc** [comment]
On further review of the ratified Zfh spec, there are fmadd.h and related instructions.  See the bottom figure of Section 15.2.


https://wiki.riscv.org/display/HOME/Recently+Ratified+Extensions
https://drive.google.com/file/d/1z3tQQLm5ALsAD77PM0l0CHnapxWCeVzP/view

**Comment by allenjbaum** [comment]
I'm not sure how to interpret that.
 "The Zfinx extension adds all of the instructions that the F extension
adds, except for the transfer instructions FLW, FSW, FMV.W.X, FMV.X.W,
C.FLW[SP], and C.FSW[SP]."

So, there should be FMA tests.

**Comment by anuani21** [comment]
Hi,

I thought fma and fmadd instructions are different. Fmadd and fmsub
instructions are available in zfh and zfinx. I will upload the coverage
report for fmadd and fmsub instructions shortly.

**Comment by davidharrishmc** [comment]
Remember that there are fnmadd and fnmsub as well in the FMA class of instructions.

**Comment by anuani21** [comment]
Ya sure I will upload fmadd,fmsub,fnmadd and fnmsub instructions.

**Comment by allenjbaum** [comment]
Ah, I thought you were using fma as the generic for fm[n][add/sub]

File: riscv-test-suite/env/arch_test.h [hunk 1]
```diff
@@ -216,12 +216,31 @@
...
+  #define FLREG ld
+  #define FSREG sd
+  #define FREGWIDTH 8
+  #define FLEN 64
+  #if XLEN==64
+    #define SIGALIGN 8
+  #else
+      #define SIGALIGN 4
+  #endif
+#elif ZDINX==1
```
Comment (allenjbaum) [inline riscv-test-suite/env/arch_test.h:238]: This looks wrong; ZFinx only needs 4B registers, NOT 8B registers. And, if both Zfinx and Zdinx exist, then you need to accommodate both sizes, don't you?  And, they're different on RV64 vs. RV32. So Zdinx would need to execute 2 separate LW ops on an RV32 compared to a single LD that RV64 to replace a FLD.

**Comment by davidharrishmc** [comment]
CORE-V Wally will take advantage of the FMA tests as soon as you have them available.

**Comment by anuani21** [comment]
I have updated FMA tests for Zfh and Zfinx extension.

**Comment by davidharrishmc** [comment]
I tried them out and they are working for me.  The f{madd/msub/nmadd/nmsub}_b15-01 tests are not in the PR.  These are the largest set of FMA tests.  Is there a reason they aren't applicable to Zfh?

**Comment by anuani21** [comment]
I will check and let you know

**Comment by anuani21** [comment]
@davidharrishmc, Are you facing any issue while running Zfinx test?

**Comment by davidharrishmc** [comment]
I don't have hardware support for Zfinx, so I don't know if those tests have any issues.

**Comment by anuani21** [comment]
No issues in the test, to check whether it is working fine.

**Comment by anuani21** [comment]
@davidharrishmc, I have updated  f{madd/msub/nmadd/nmsub}_b15-01 tests in the PR.

**Comment by davidharrishmc** [comment]
The fcvt.h.l tests should only depend on the F and Zfh extensions, as indicated in the test case.

```
RVTEST_CASE(0,"//check ISA:=regex(.*64.*I.*F.*Zfh.*);def TEST_CASE_1=True;",fcvt.h.l_b25)
inst_0:// rs1==x31, rd==f31,rs1_val == 0 and  fcsr == 0 and rm_val == 7  
/* opcode: fcvt.h.l ; op1:x31; dest:f31; op1val:0x0; valaddr_reg:x3;
val_offset:0*8; rmval:dyn; correctval:??; testreg:x2;
fcsr_val: 0*/
TEST_FPIO_OP(fcvt.h.l, f31, x31, dyn, 0, 0, x3, 0*8, x4, x1, x2,ld)
```

However, the TEST_FPIO_OP macro generates a fsd, which requires the D extension.  Therefore, these tests do not run successfully on a RV64IFZfh machine.  

```
0000000080000440 <inst_0>:
    80000440:	0001bf83          	ld	t6,0(gp)
    80000444:	00000113          	li	sp,0
    80000448:	00311073          	fscsr	sp
    8000044c:	d42fffd3          	fcvt.h.l	ft11,t6
    80000450:	00302273          	frcsr	tp
    80000454:	01f0b027          	fsd	ft11,0(ra)
    80000458:	0040b423          	sd	tp,8(ra)
```

It would be better to produce a fsw or fsh in this situation.

**Comment by allenjbaum** [comment]
I see no further commits that would indicate that the problem David found (tests use D-ext ops on a configuration that only supports F-extension) have been fixed. Has that one been fixed?

As far as I can tell, that "fsd" corresponds to the first parameter of TEST_FPIO_OP ("inst" parameter ) 
but I can't find any instance of "fsd* used anywhere in the repo, so I don't have any idea where it is coming from.

**Comment by allenjbaum** [comment]
Actually, that "fsd" op corresponds to something after the FPIO macro, but all the FP tests that use the FPIO macro consist of nothing but FPIO macros. the only ops that use FPIO are FCVT ops, and there are no uses of fsd in them, therefore it must be after the FPIO. But, all of those tests end with no further ops. Search for "fsd" shows only tests of fsd, and the don't use FPIO. So, I cannot figure out where that fsd is being generated.

**Comment by anuani21** [comment]
@david and @allen ,I will check and get back to you. Sorry for the delay.

**Comment by anuani21** [comment]
@davidharrishmc  and @allenjbaum,

I have updated the test using fsw instead of fsd. I have attached the logs for your reference for fcvt.h.l instruction.

00000000800003ac <inst_0>:
    800003ac:	0001bf83          	ld	t6,0(gp)
    800003b0:	00000113          	li	sp,0
    800003b4:	00311073          	fscsr	sp
    800003b8:	d42fffd3          	fcvt.h.l	ft11,t6
    800003bc:	00302273          	frcsr	tp
    800003c0:	01f0a027          	fsw	ft11,0(ra)
    800003c4:	0040b223          	sd	tp,4(ra)

**Comment by allenjbaum** [comment]
Would you mind submitting a PR for the bugs you found?

**Comment by davidharrishmc** [comment]
Zfh isn’t in the repo yet, so I don’t know how to do a PR on a PR.  Could you either accept PR367 and I’ll do a PR with these fixes, or anuani could incorporate fixes into PR367?

David

**Comment by allenjbaum** [comment]
Oh, and anuani21 - thanks for figuring out what the problem is. 
Now we know we should carefully review this macros as well as the tests.
It is still a puzzle to me that this wasn't caught ages ago.

**Review by allenjbaum — CHANGES_REQUESTED** [review]
This looks good, and thanks for figuring out what the problem was. I would ask that you back out the changes in the comments however, as the clutter up the diff. We can merge it as soon as that is done.

**Comment by allenjbaum** [comment]
yes. I want some changes to the PR #367 first though, to back out changes
to the comments.
Those are unnecessary, and clutter up the diff - and will clutter the diff
up again if someone regenerates the tests, so it should ust be left
along (and, yes, I know I can filter out comments fro the diff...)

**Review by allenjbaum — APPROVED** [review]
This unneccessarily (in my opinion, anyway) edits comments - but I don't think that will cause future problem. It also has a simple fix for the issues that were seen. We may want further tests in the future to more carefully look at corner cases for better coverage, but this is at a minimum a good start

**Comment by allenjbaum** [comment]
sigh: as usual, we have a conflict with the changelog version. It needs to be updated.
Note that in future, all PRs should be filed against the dev branch, the dev branch should not be checking for changelog version numbers, and maintainers will be responsible for updating the union of changelogs for all the PRs included in that version of the dev branch

**Comment by anuani21** [comment]
@allenjbaum, I have updated the changelog file. Please have your review and let us know if more information would help.

Hereafter, I will raise the PR with target branch as dev.

**Review by allenjbaum — APPROVED** [review]
This fixes the problem, and the previous fixes the earlier problem of specifying D support for an extension that doesn't need it.

**Comment by stnolting** [comment]
Just one question about this: `Zfinx` is currently only supported for `rv64` machines, right?! 🤔

(there is a `rv64i_m/Zfinx` testcases folder but no `rv32i_m/Zfinx` testcases folder)
